rpmalloc/rpmalloc.c

```diff
@@ -2380,8 +2380,22 @@ _rpmalloc_deallocate_direct_small_or_medium(span_t* span, void* block) {
 	--span->used_count;
 	span->free_list = block;
 	if (UNEXPECTED(span->used_count == span->list_size)) {
-		_rpmalloc_span_double_link_list_remove(&heap->size_class[span->size_class].partial_span, span);
-		_rpmalloc_span_release_to_cache(heap, span);
+		int all_remaining_free_or_deferred = 1;
+		// If there are no used blocks it is guaranteed that no other external thread is accessing the span
+		if (span->used_count) {
+			// Make sure we have synchronized the deferred list and list size by using acquire semantics
+			// and guarantee that no external thread is accessing span concurrently
+			void* free_list;
+			do {
+				free_list = atomic_exchange_ptr_acquire(&span->free_list_deferred, INVALID_POINTER);
+			} while (free_list == INVALID_POINTER);
+			all_remaining_free_or_deferred = (span->used_count == span->list_size);
```

Comment:
As the check for `span->used_count == span->list_size` has already been satisfied above, assuming "correct" usage, wouldn't this second check always be `true`? IIUC, `span->list_size <= span->used_count` should always be `true` and any other threads can only increment `span->list_size`. But perhaps I'm missing something.

Reply:
Hmm, yeah, that is correct, we just need the synchronization on free_list_deferred to ensure any external thread has finished adding to the deferred list. The second check of span->used_count == span->list_size is redundant